dynamiqs/smesolve/smediffrax.py

```diff
@@ -0,0 +1,117 @@
+from jaxtyping import PyTree
+
+from ..core.abstract_solver import SMESolver
+from ..core.diffrax_solver import AdaptiveSolver, DiffraxSolver, EulerSolver
+from ..result import Saved, SMESaved
+from ..utils.utils.general import dag, trace, tracemm
+
+
+# state for the diffrax solver for SMEs
+class Y(eqx.Module):
+    rho: Array
+    dYt: Array
+
+
+class MeasurementTerm(dx.ControlTerm):
+    def prod(self, vf: dx.VF, control: dx.Control) -> dx.Y:
+        dW = control
+        rho = (vf.rho * dW[:, None, None]).sum(0)  # (n, n)
+        return Y(rho, dW)
+
+
+class SMEDiffraxSolver(DiffraxSolver, SMESolver):
+    wiener: dx.VirtualBrownianTree
+
+    def __init__(self, *args):
+        # === pass all init arguments to `BaseSolver`
+        super().__init__(*args)
+
+        # === define save function to save measurement results
+        fn = lambda t, y, args: y.dYt  # noqa: ARG005
+        save_c = dx.SubSaveAt(ts=self.tmeas, fn=fn)  # save measurement results
+        self.saveat.subs.append(save_c)
+
+        # === define initial augmented state
+        self.y0 = Y(self.y0, jnp.empty(len(self.etas)))
+
+        # === define wiener process
+        self.wiener = dx.VirtualBrownianTree(
+            self.t0, self.t1, tol=1e-3, shape=(len(self.etas),), key=self.key
+        )  # todo: fix hard-coded tol
+
+    @property
+    def terms(self) -> dx.AbstractTerm:
+        # === define deterministic term
+        # This contains everything before the "dt" in the SME:
+        # - lindblad term for drho
+        # - sqrt(eta) Tr[(L+Ld) @ rho] for dYt
+        def vector_field(t, y, _):  # noqa: ANN001, ANN202
+            # state
+            Ls = jnp.stack([L(t) for L in self.Ls])
+            Lsd = dag(Ls)
+            LdL = (Lsd @ Ls).sum(0)
+            H = self.H(t)
+            tmp = (-1j * H - 0.5 * LdL) @ y.rho + 0.5 * (Ls @ y.rho @ Lsd).sum(0)
+            rho = tmp + dag(tmp)
+
+            # signal
+            Lms = jnp.stack([L(t) for L in self.Lms])
+            tr_Lms_rho = tracemm(Lms, y.rho)
+            dYt = jnp.sqrt(self.etas) * (tr_Lms_rho + tr_Lms_rho.conj()).real  # (nLm,)
+
+            return Y(rho, dYt)
+
+        lindblad_term = dx.ODETerm(vector_field)
+
+        # === define stochastic term
+        # This contains everything before the "dWt" in the SME:
+        # - measurement backaction term for drho
+        # - simply dWt for dYt
+        def vector_field(t, y, _):  # noqa: ANN001, ANN202
+            # \sqrt\eta (L @ rho + rho @ Ld - Tr[L @ rho + rho @ Ld] rho) dWt
+            Lms = jnp.stack([L(t) for L in self.Lms])
+            Lms_rho = Lms @ y.rho
+            tmp = Lms_rho + dag(Lms_rho)
+            tr = trace(tmp).real
+
+            # state
+            etas = self.etas[:, None, None]  # (nLm, n, n)
+            tr = tr[:, None, None]  # (nLm, n, n)
+            rho = jnp.sqrt(etas) * (tmp - tr * y.rho)  # (nLm, n, n)
+
+            # signal
+            dYt = jnp.empty(len(self.etas))
```

Comment:
Should this not be `jnp.zeros` instead? `empty` will just allocate memory without initializing, which might be added to the actual result later on. Maybe I'm missing something.

Reply:
This is correct. In `prod` we just throw this to the bin, and return `Y(rho, dW)` (there is no signal-dependent term in the stochastic part, it's just `dYt = deterministic + dWt`. I think using `jnp.empty` is a good way to express that this is unused, I can maybe put a comment to make it more clear.

Reply:
Another option (to explain) would be to do `dYt = jnp.ones(len(self.etas))` and in `prod`:
```
    def prod(self, vf: dx.VF, control: dx.Control) -> dx.Y:
        dW = control
        rho = (vf.rho * dW[:, None, None]).sum(0)  # (n, n)
        dY = vf.dY * dW
        return Y(rho, dY)
```